streams/src/main/java/org/apache/kafka/streams/processor/assignment/TaskAssignmentUtils.java

```diff
@@ -244,18 +274,117 @@ public static Map<ProcessId, KafkaStreamsAssignment> optimizeRackAwareStandbyTas
         );
         LOG.info("Assignment before standby task optimization has cost {}", initialCost);
 
-        throw new UnsupportedOperationException("Not yet Implemented.");
+        final MoveStandbyTaskPredicate moveablePredicate = getStandbyTaskMovePredicate(applicationState);
+        final BiFunction<KafkaStreamsAssignment, KafkaStreamsAssignment, List<TaskId>> getMovableTasks = (source, destination) -> {
+            return source.tasks().values().stream().filter(task -> task.type() == AssignedTask.Type.STANDBY)
+                .filter(task -> {
+                    return !destination.tasks().containsKey(task.id());
+                })
+                .filter(task -> {
+                    final KafkaStreamsState sourceState = kafkaStreamsStates.get(source.processId());
+                    final KafkaStreamsState destinationState = kafkaStreamsStates.get(source.processId());
+                    return moveablePredicate.canMoveStandbyTask(sourceState, destinationState, task.id(), kafkaStreamsAssignments);
+                })
+                .map(AssignedTask::id)
+                .sorted()
+                .collect(Collectors.toList());
+        };
+
+        final long startTime = System.currentTimeMillis();
+        boolean taskMoved = true;
+        int round = 0;
+        final RackAwareGraphConstructor<KafkaStreamsAssignment> graphConstructor = RackAwareGraphConstructorFactory.create(
+            applicationState.assignmentConfigs().rackAwareAssignmentStrategy(), taskIds);
+        while (taskMoved && round < STANDBY_OPTIMIZER_MAX_ITERATION) {
+            taskMoved = false;
+            round++;
+            for (int i = 0; i < kafkaStreamsAssignments.size(); i++) {
+                final UUID clientId1 = clientIds.get(i);
+                final KafkaStreamsAssignment clientState1 = kafkaStreamsAssignments.get(new ProcessId(clientId1));
+                for (int j = i + 1; j < kafkaStreamsAssignments.size(); j++) {
+                    final UUID clientId2 = clientIds.get(i);
+                    final KafkaStreamsAssignment clientState2 = kafkaStreamsAssignments.get(new ProcessId(clientId2));
+
+                    final String rack1 = clientRacks.get(clientState1.processId().id()).get();
+                    final String rack2 = clientRacks.get(clientState2.processId().id()).get();
```

Comment:
Future cleanup: we can actually get rid of this clientRacks data structure altogether, and just use the rack ids we already processed and nicely organized in the KafkaStreamsState which is addressable by UUID, ie

```suggestion
final String rack1 = kafkaStreamsStates.get(clientState1.processId().id()).rackId().get();
final String rack2 = kafkaStreamsStates.get(clientState2.processId().id()).rackId().get();
```